scipy/stats/_stats_py.py

```diff
@@ -100,13 +100,18 @@ def _contains_nan(a, nan_policy='propagate'):
         try:
             contains_nan = np.any(np.isnan(a))
         except TypeError:
-            # Don't know what to do. Fall back to omitting nan values and
-            # issue a warning.
-            contains_nan = False
-            nan_policy = 'omit'
-            warnings.warn("The input array could not be properly "
-                          "checked for nan values. nan values "
-                          "will be ignored.", RuntimeWarning)
+            try:
+                # This can happen when attempting to check nan with np.isnan
+                # for string array (e.g. as in the function `rankdata`).
+                contains_nan = np.any(a == "nan")
```

Comment:
There are two remaining points of criticism I can think of, but neither bothers me:
* it looks like this will _only_ find `"nan"` string arrays and would not find a true `nan`, e.g. in an object array. 
* From a different perspective, it's arguable that the string "nan" should be treated like any other string rather than a Not a Number, since _every_ string is not a number. 

@tupui do you have an opinion here? We're running into the same trouble here that led to the deprecation of non-numerical input to `mode` in gh-15573. Ultimately, I think we might want to do the same for `rankdata`: deprecate and point people to Pandas for ranking non-numerical data. So I wouldn't mind merging this as-is.

Reply:
I am fine with this PR getting in as it's a clear improvement.

But yes, ultimately we should either ask people to user Pandas to pre-process their data (I prefer that, clearer separation of responsibilities. E.g. to draw a parallel on Pandas' side, to me you don't want to use Pandas for stats and plotting even though they do this for conveniency) or we will have to talk with Pandas to have similar behaviour. And then maybe this calls for a common nan policy library (why not).

Reply:
@AtsushiSakai @tupui I hate to suggest it, but what are your thoughts on AtsushiSakai/scipy#1 - an explicit loop to check for true `np.nan`s? I'm less comfortable with considering "nan" to be a NaN value.